In [7]:
# Imports
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [21]:
# Load in data and display first 5 rows
df = pd.read_csv('Life Expectancy Data.csv')
df.head()

Country  Year      Status  Life expectancy   Adult Mortality  \
0  Afghanistan  2015  Developing              65.0            263.0   
1  Afghanistan  2014  Developing              59.9            271.0   
2  Afghanistan  2013  Developing              59.9            268.0   
3  Afghanistan  2012  Developing              59.5            272.0   
4  Afghanistan  2011  Developing              59.2            275.0   

   infant deaths  Alcohol  percentage expenditure  Hepatitis B  Measles   ...  \
0             62     0.01               71.279624         65.0      1154  ...   
1             64     0.01               73.523582         62.0       492  ...   
2             66     0.01               73.219243         64.0       430  ...   
3             69     0.01               78.184215         67.0      2787  ...   
4             71     0.01                7.097109         68.0      3013  ...   

   Polio  Total expenditure  Diphtheria    HIV/AIDS         GDP  Population  \
0    6.0               8.16         65.0        0.1  584.259210  33736494.0   
1   58.0               8.18         62.0        0.1  612.696514    327582.0   
2   62.0               8.13         64.0        0.1  631.744976  31731688.0   
3   67.0               8.52         67.0        0.1  669.959000   3696958.0   
4   68.0               7.87         68.0        0.1   63.537231   2978599.0   

    thinness  1-19 years   thinness 5-9 years  \
0                   17.2                 17.3   
1                   17.5                 17.5   
2                   17.7                 17.7   
3                   17.9                 18.0   
4                   18.2                 18.2   

   Income composition of resources  Schooling  
0                            0.479       10.1  
1                            0.476       10.0  
2                            0.470        9.9  
3                            0.463        9.8  
4                            0.454        9.5  

[5 rows x 22 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2938 entries, 0 to 2937
Data columns (total 22 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   Country                          2938 non-null   object 
 1   Year                             2938 non-null   int64  
 2   Status                           2938 non-null   object 
 3   Life expectancy                  2928 non-null   float64
 4   Adult Mortality                  2928 non-null   float64
 5   infant deaths                    2938 non-null   int64  
 6   Alcohol                          2744 non-null   float64
 7   percentage expenditure           2938 non-null   float64
 8   Hepatitis B                      2385 non-null   float64
 9   Measles                          2938 non-null   int64  
 10   BMI                             2904 non-null   float64
 11  under-five deaths                2938 non-null   int64  
 12  Polio               

- 2938 rows and 22 columns
- Mostly numerical data with Country and Status listed as strings
- Null values in several columns. Highest number in Populations, GDP, Hepatitis, Income composition of resources, and Schooling

In [4]:
df.isna().sum()

Country                              0
Year                                 0
Status                               0
Life expectancy                     10
Adult Mortality                     10
infant deaths                        0
Alcohol                            194
percentage expenditure               0
Hepatitis B                        553
Measles                              0
 BMI                                34
under-five deaths                    0
Polio                               19
Total expenditure                  226
Diphtheria                          19
 HIV/AIDS                            0
GDP                                448
Population                         652
 thinness  1-19 years               34
 thinness 5-9 years                 34
Income composition of resources    167
Schooling                          163
dtype: int64

Possible routes of exploration:
1. Disease trends by year (Regions vs Global)
    a. Percentage expenditure vs disease trends and GDP per country
    b. Income composition (does higher quality of life == less disease?) vs disease trends
2. Is there a relationship between Schooling and diseases / life expectancy?
3. GDP vs life expectancy
    a. GDP per capita?

*Group by continent, status

In [13]:
relevant_cols = [
    'Country',
    'Year',
    'Status',
    'Life expectancy',
    'Adult Mortality',
    'under-five deaths',
    'percentage expenditure',
    'GDP',
    'Population',
    'Income composition of resources',
    'Schooling'
]

In [14]:
df_relevant = df[relevant_cols]

KeyError: "['Life expectancy', 'under-five deaths'] not in index"

In [17]:
df.columns

Index(['Country', 'Year', 'Status', 'Life expectancy ', 'Adult Mortality',
       'infant deaths', 'Alcohol', 'percentage expenditure', 'Hepatitis B',
       'Measles ', ' BMI ', 'under-five deaths ', 'Polio', 'Total expenditure',
       'Diphtheria ', ' HIV/AIDS', 'GDP', 'Population',
       ' thinness  1-19 years', ' thinness 5-9 years',
       'Income composition of resources', 'Schooling'],
      dtype='object')

In [16]:
df.columns.str.strip()

Index(['Country', 'Year', 'Status', 'Life expectancy', 'Adult Mortality',
       'infant deaths', 'Alcohol', 'percentage expenditure', 'Hepatitis B',
       'Measles', 'BMI', 'under-five deaths', 'Polio', 'Total expenditure',
       'Diphtheria', 'HIV/AIDS', 'GDP', 'Population', 'thinness  1-19 years',
       'thinness 5-9 years', 'Income composition of resources', 'Schooling'],
      dtype='object')

In [22]:
df = df.rename(columns={'Life expectancy ':'Life expectancy', 'under-five deaths ':'under-five deaths'})
df.columns

Index(['Country', 'Year', 'Status', 'Life expectancy', 'Adult Mortality',
       'infant deaths', 'Alcohol', 'percentage expenditure', 'Hepatitis B',
       'Measles ', ' BMI ', 'under-five deaths', 'Polio', 'Total expenditure',
       'Diphtheria ', ' HIV/AIDS', 'GDP', 'Population',
       ' thinness  1-19 years', ' thinness 5-9 years',
       'Income composition of resources', 'Schooling'],
      dtype='object')

In [24]:
df_relevant = df[relevant_cols]
df_relevant

Country  Year      Status  Life expectancy  Adult Mortality  \
0     Afghanistan  2015  Developing             65.0            263.0   
1     Afghanistan  2014  Developing             59.9            271.0   
2     Afghanistan  2013  Developing             59.9            268.0   
3     Afghanistan  2012  Developing             59.5            272.0   
4     Afghanistan  2011  Developing             59.2            275.0   
...           ...   ...         ...              ...              ...   
2933     Zimbabwe  2004  Developing             44.3            723.0   
2934     Zimbabwe  2003  Developing             44.5            715.0   
2935     Zimbabwe  2002  Developing             44.8             73.0   
2936     Zimbabwe  2001  Developing             45.3            686.0   
2937     Zimbabwe  2000  Developing             46.0            665.0   

      under-five deaths  percentage expenditure         GDP  Population  \
0                    83               71.279624  584.259210  33736494.0   
1                    86               73.523582  612.696514    327582.0   
2                    89               73.219243  631.744976  31731688.0   
3                    93               78.184215  669.959000   3696958.0   
4                    97                7.097109   63.537231   2978599.0   
...                 ...                     ...         ...         ...   
2933                 42                0.000000  454.366654  12777511.0   
2934                 41                0.000000  453.351155  12633897.0   
2935                 40                0.000000   57.348340    125525.0   
2936                 39                0.000000  548.587312  12366165.0   
2937                 39                0.000000  547.358878  12222251.0   

      Income composition of resources  Schooling  
0                               0.479       10.1  
1                               0.476       10.0  
2                               0.470        9.9  
3                               0.463        9.8  
4                               0.454        9.5  
...                               ...        ...  
2933                            0.407        9.2  
2934                            0.418        9.5  
2935                            0.427       10.0  
2936                            0.427        9.8  
2937                            0.434        9.8  

[2938 rows x 11 columns]

In [25]:
df_relevant.describe()

Year  Life expectancy  Adult Mortality  under-five deaths  \
count  2938.000000      2928.000000      2928.000000        2938.000000   
mean   2007.518720        69.224932       164.796448          42.035739   
std       4.613841         9.523867       124.292079         160.445548   
min    2000.000000        36.300000         1.000000           0.000000   
25%    2004.000000        63.100000        74.000000           0.000000   
50%    2008.000000        72.100000       144.000000           4.000000   
75%    2012.000000        75.700000       228.000000          28.000000   
max    2015.000000        89.000000       723.000000        2500.000000   

       percentage expenditure            GDP    Population  \
count             2938.000000    2490.000000  2.286000e+03   
mean               738.251295    7483.158469  1.275338e+07   
std               1987.914858   14270.169342  6.101210e+07   
min                  0.000000       1.681350  3.400000e+01   
25%                  4.685343     463.935626  1.957932e+05   
50%                 64.912906    1766.947595  1.386542e+06   
75%                441.534144    5910.806335  7.420359e+06   
max              19479.911610  119172.741800  1.293859e+09   

       Income composition of resources    Schooling  
count                      2771.000000  2775.000000  
mean                          0.627551    11.992793  
std                           0.210904     3.358920  
min                           0.000000     0.000000  
25%                           0.493000    10.100000  
50%                           0.677000    12.300000  
75%                           0.779000    14.300000  
max                           0.948000    20.700000

- Wide range between min/max for Life expetancy
- Scientific notation in population
- under-five deaths mean is highly skewed upward (mean = 42; median = 4)
- At least one country that has 0 years of schooling

In [33]:
df_relevant.describe(include='object')

Country      Status
count          2938        2938
unique          193           2
top     Afghanistan  Developing
freq             16        2426

In [35]:
df_relevant['Status'].value_counts()

Status
Developing    2426
Developed      512
Name: count, dtype: int64

- 193 unique countries with statuses: 'Developed' or 'Developing'

In [39]:
# Feature Engineer column to group countries by Region
import json
with open('country-by-continent.json') as file:
    country_continent_converter = json.load(file)

In [40]:
print(country_continent_converter)

[{'country': 'Afghanistan', 'continent': 'Asia'}, {'country': 'Albania', 'continent': 'Europe'}, {'country': 'Algeria', 'continent': 'Africa'}, {'country': 'American Samoa', 'continent': 'Oceania'}, {'country': 'Andorra', 'continent': 'Europe'}, {'country': 'Angola', 'continent': 'Africa'}, {'country': 'Anguilla', 'continent': 'North America'}, {'country': 'Antarctica', 'continent': 'Antarctica'}, {'country': 'Antigua and Barbuda', 'continent': 'North America'}, {'country': 'Argentina', 'continent': 'South America'}, {'country': 'Armenia', 'continent': 'Asia'}, {'country': 'Aruba', 'continent': 'North America'}, {'country': 'Australia', 'continent': 'Oceania'}, {'country': 'Austria', 'continent': 'Europe'}, {'country': 'Azerbaijan', 'continent': 'Asia'}, {'country': 'Bahamas', 'continent': 'North America'}, {'country': 'Bahrain', 'continent': 'Asia'}, {'country': 'Bangladesh', 'continent': 'Asia'}, {'country': 'Barbados', 'continent': 'North America'}, {'country': 'Belarus', 'continent

In [41]:
type(country_continent_converter)

list

In [42]:
country_continent_converter[0]

{'country': 'Afghanistan', 'continent': 'Asia'}

In [43]:
country_continent_converter[0]["country"]

'Afghanistan'

In [44]:
country_continent_converter[0]["continent"]

'Asia'

In [46]:
for country_continent_dict in country_continent_converter:
    if 'Afghanistan' == country_continent_dict["country"]:
        print(country_continent_dict["continent"])


Asia


In [48]:
for country in df_relevant['Country']:
    for country_continent_dict in country_continent_converter:
        if country_continent_dict == country_continent_dict["country"]:
            print(country_continent_dict["continent"])

In [56]:
df_relevant['continent'] = [continent_dict["continent"] for continent_dict in country_continent_converter if df_relevant['Country'] == continent_dict["country"]][0]

ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().

In [57]:
# We can write a function for this
def fixed_mapper(country_from_df):
    '''
    Takes in the current status of animals and outputs whether they have been fixed
    '''
    for country_continent_dict in country_continent_converter:
        if country_from_df == country_continent_dict["country"]:
            return country_continent_dict["continent"]
        
    # if country_from_df loops through converter and doesn't find a continent
    # then return "N/A"
    return "N/A"
    

In [74]:
# def get_continent(country): 
#        return [continent_dict["continent"] for continent_dict in country_continent_converter if continent_dict["country"] == country] 


# df_relevant['continent_1'] = df_relevant['Country'].apply(get_continent)

C:\Users\Elija\AppData\Local\Temp\ipykernel_27748\1183714501.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_relevant['continent_1'] = df_relevant['Country'].apply(get_continent)


In [58]:
df_relevant["continent"] = df_relevant["Country"].map(fixed_mapper)

C:\Users\Elija\AppData\Local\Temp\ipykernel_27748\3065947453.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_relevant["continent"] = df_relevant["Country"].map(fixed_mapper)


In [75]:
df_relevant.head(10)

Country  Year      Status  Life expectancy  Adult Mortality  \
0  Afghanistan  2015  Developing             65.0            263.0   
1  Afghanistan  2014  Developing             59.9            271.0   
2  Afghanistan  2013  Developing             59.9            268.0   
3  Afghanistan  2012  Developing             59.5            272.0   
4  Afghanistan  2011  Developing             59.2            275.0   
5  Afghanistan  2010  Developing             58.8            279.0   
6  Afghanistan  2009  Developing             58.6            281.0   
7  Afghanistan  2008  Developing             58.1            287.0   
8  Afghanistan  2007  Developing             57.5            295.0   
9  Afghanistan  2006  Developing             57.3            295.0   

   under-five deaths  percentage expenditure         GDP  Population  \
0                 83               71.279624  584.259210  33736494.0   
1                 86               73.523582  612.696514    327582.0   
2                 89               73.219243  631.744976  31731688.0   
3                 93               78.184215  669.959000   3696958.0   
4                 97                7.097109   63.537231   2978599.0   
5                102               79.679367  553.328940   2883167.0   
6                106               56.762217  445.893298    284331.0   
7                110               25.873925  373.361116   2729431.0   
8                113               10.910156  369.835796  26616792.0   
9                116               17.171518  272.563770   2589345.0   

   Income composition of resources  Schooling continent continent_1  
0                            0.479       10.1      Asia      [Asia]  
1                            0.476       10.0      Asia      [Asia]  
2                            0.470        9.9      Asia      [Asia]  
3                            0.463        9.8      Asia      [Asia]  
4                            0.454        9.5      Asia      [Asia]  
5                            0.448        9.2      Asia      [Asia]  
6                            0.434        8.9      Asia      [Asia]  
7                            0.433        8.7      Asia      [Asia]  
8                            0.415        8.4      Asia      [Asia]  
9                            0.405        8.1      Asia      [Asia]

In [76]:
df_relevant['continent'].value_counts()

continent
Africa           784
Asia             624
Europe           546
N/A              368
North America    322
South America    160
Oceania          134
Name: count, dtype: int64

In [82]:
test = df_relevant.loc[df_relevant['continent'] == 'N/A']
test['Country'].unique()

array(['Bolivia (Plurinational State of)', 'Brunei Darussalam',
       "Côte d'Ivoire", 'Cabo Verde', 'Czechia',
       "Democratic People's Republic of Korea",
       'Democratic Republic of the Congo', 'Fiji',
       'Iran (Islamic Republic of)', "Lao People's Democratic Republic",
       'Micronesia (Federated States of)', 'Republic of Korea',
       'Republic of Moldova', 'Russian Federation', 'Swaziland',
       'Syrian Arab Republic',
       'The former Yugoslav republic of Macedonia', 'Timor-Leste',
       'United Kingdom of Great Britain and Northern Ireland',
       'United Republic of Tanzania', 'United States of America',
       'Venezuela (Bolivarian Republic of)', 'Viet Nam'], dtype=object)

In [83]:
relabel_countries = {
    'Bolivia (Plurinational State of)': 'Bolivia',
    'Brunei Darussalam': 'Brunei',
       "Côte d'Ivoire": 'Ivory Coast',
       'Cabo Verde': 'Cape Verde',
       'Czechia': 'Czech Republic',
       "Democratic People's Republic of Korea": 'North Korea',
       'Democratic Republic of the Congo': 'The Democratic Republic of Congo',
       'Fiji': 'Fiji Islands',
       'Iran (Islamic Republic of)': 'Iran',
       "Lao People's Democratic Republic": 'Laos',
       'Micronesia (Federated States of)': 'Micronesia, Federated States of',
       'Republic of Korea': 'South Korea',
       'Republic of Moldova': 'Moldova',
       'Russian Federation': 'Russia',
       'Swaziland': 'Eswatini',
       'Syrian Arab Republic': 'Syria',
       'The former Yugoslav republic of Macedonia': 'North Macedonia',
       'Timor-Leste': 'East Timor',
       'United Kingdom of Great Britain and Northern Ireland': 'United Kingdom',
       'United Republic of Tanzania': 'Tanzania',
       'United States of America': 'United States',
       'Venezuela (Bolivarian Republic of)': 'Venezuela',
       'Viet Nam': 'Vietnam'
}

In [86]:
df_relevant['Country'] = df_relevant['Country'].replace(relabel_countries)

C:\Users\Elija\AppData\Local\Temp\ipykernel_27748\852530127.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_relevant['Country'] = df_relevant['Country'].replace(relabel_countries)


In [85]:
test = df_relevant.loc[df_relevant['continent'] == 'N/A']
test['Country'].unique()

array(['Bolivia', 'Brunei', 'Ivory Coast', 'Cape Verde', 'Czech Republic',
       'North Korea', 'The Democratic Republic of Congo', 'Fiji Islands',
       'Iran', 'Laos', 'Micronesia, Federated States of', 'South Korea',
       'Moldova', 'Russia', 'Eswatini', 'Syria', 'North Macedonia',
       'East Timor', 'United Kingdom', 'Tanzania', 'United States',
       'Venezuela', 'Vietnam'], dtype=object)

In [88]:
df_relevant["continent"] = df_relevant["Country"].map(fixed_mapper)

C:\Users\Elija\AppData\Local\Temp\ipykernel_27748\3065947453.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_relevant["continent"] = df_relevant["Country"].map(fixed_mapper)


In [89]:
test = df_relevant.loc[df_relevant['continent'] == 'N/A']
test['Country'].unique()

array([], dtype=object)